File: matlab/examples/QuickStart.m
% Basic VAR simulation
vrng = varmapack.Rng(123);
cleanup = onCleanup(@() delete(vrng));
n = 200;
M = 10;
Sig1 = [2, 0; 0, 1];
A = [0.6, 0.1; 0, 0.4];
X = varmapack.sim(A, [], Sig1, [], n, 1, [], vrng);
fprintf("VAR simulation, first five values:\n");
disp(X(:, 1:5))

% Named testcases
fprintf("\nNamed testcases:\n");
varmapack.testcase('summary')
[A, B, Sig] = varmapack.testcase('smallARMA1');
X = varmapack.sim(A, B, Sig, [], n, M, [], vrng);
fprintf("smallARMA1 simulation size: %s\n", mat2str(size(X)));

% Time-dependent means and fixed startup paths
mu = [0, 1; 0, 1];
X0 = [0, 0.1; 0, -0.1];
X0multi = cat(3, X0, X0 + [0.05, 0.05; -0.05, -0.05]);
Xcommon = varmapack.sim(A, B, Sig, mu, n, M, X0, vrng);
Xseparate = varmapack.sim(A, B, Sig, mu, n, 2, X0multi, vrng);
fprintf("fixed-start simulation sizes: %s and %s\n", ...
        mat2str(size(Xcommon)), mat2str(size(Xseparate)));

% VAR, VMA, and VARMA simulation
A1 = [0.6, 0.1; 0, 0.4];
B1 = [0.4, 0.3; 0.1, 0.2];
A2 = 0.2*ones(2);
B2 = 0.5*eye(2);
X1 = varmapack.sim(A1, [], Sig, [], n, M, [], vrng);
X2 = varmapack.sim([], [B1, B2], Sig, [], n, M, [], vrng);
[X3, E] = varmapack.sim([A1, A2], B1, Sig, [], n, M, [], vrng);
fprintf("VAR, VMA, VARMA, and shock sizes: %s, %s, %s, %s\n", ...
        mat2str(size(X1)), mat2str(size(X2)), mat2str(size(X3)), ...
        mat2str(size(E)));

% VARMAX simulation with common and replicate-specific vector exogenous inputs
C = [0.8, -0.3; 0.2, 0.4];
z = [cos((1:n)/10); sin((1:n)/10)];
zmulti = cat(3, z, z + 0.01);
[X4, E4] = varmapack.simx(A1, B1, C, z, Sig, n, M, zeros(2), 2, vrng);
X5 = varmapack.simx(A1, B1, C, zmulti, Sig, n, 2, zeros(2), 2, vrng);
fprintf("VARMAX and shock sizes: %s, %s, %s\n", ...
        mat2str(size(X4)), mat2str(size(E4)), mat2str(size(X5)));

% Model analysis
maxlag = 3;
rho = varmapack.specrad([A1, A2]);
rhoMA = varmapack.ma_specrad(B1);
Gamma = varmapack.acvf([A1, A2], B1, Sig, maxlag);
GammaHat = varmapack.autocov(X3(:, :, 1), maxlag, "ML");
GammaHatC = varmapack.autocov(X3(:, :, 1), maxlag, "C");
Theta = varmapack.irf([A1, A2], B1, Sig, maxlag);
Psi = varmapack.psi([A1, A2], B1, maxlag);
fprintf("\nVARMA model analysis:\n");
fprintf("AR spectral radius: %.6f\n", rho);
fprintf("MA spectral radius: %.6f\n", rhoMA);
fprintf("Psi, Theta, and covariance sizes: %s, %s, %s, %s, %s\n", ...
        mat2str(size(Psi)), mat2str(size(Theta)), mat2str(size(Gamma)), ...
        mat2str(size(GammaHat)), mat2str(size(GammaHatC)));

File: python/examples/quickstart.py
import randompack
import numpy as np
import varmapack

# Basic VAR simulation
Sig = [[2, 0], [0, 1]]
VAR_model = varmapack.Model(A=[[0.6, 0.1], [0.0, 0.4]], Sig=Sig)
X = VAR_model.sim(200)
print("VAR simulation, first five values of its first replicate:")
print(X[0, :5])

# Named testcases
print("\nNamed testcases:")
print(varmapack.testcases())
model = varmapack.testcase("smallARMA1")
X = model.sim(200, nrep=10)
print("smallARMA1 simulation shape:", X.shape)

# VAR, VMA, VARMA, and VARMAX simulation
A1 = np.array([[0.6, 0.1], [0.0, 0.4]])
B1 = np.array([[0.4, 0.3], [0.1, 0.2]])
A2 = 0.2*np.ones((2, 2))
B2 = 0.5*np.eye(2)
z = np.column_stack((np.cos(np.arange(200)/10), np.sin(np.arange(200)/10)))
VARMA_model = varmapack.Model(A=[A1, A2], B=B1, Sig=Sig)
VMA_model = varmapack.Model(B=[B1, B2], Sig=Sig)
VARMAX_model = varmapack.Model(A=A1, B=B1,
                               C=[[[0.8, 0.2], [-0.3, 0.4]]], Sig=Sig)
rng = randompack.Rng()
rng.seed(123)
X1 = VAR_model.sim(200, nrep=100)
X2 = VMA_model.sim(200, nrep=100, rng=rng)
X3, E = VARMA_model.sim(200, nrep=100, rng=rng, return_shocks=True)
X4 = VARMAX_model.sim(200, nrep=100, X0=np.zeros((2, 2)), z=z, rng=rng)
print("\nSimulation shapes:")
print("VAR:   ", X1.shape)
print("VMA:   ", X2.shape)
print("VARMA: ", X3.shape, "with shocks", E.shape)
print("VARMAX:", X4.shape)

# Model analysis
maxlag = 3
rho = VARMA_model.specrad()
rhoMA = VARMA_model.ma_specrad()
Psi = VARMA_model.psi(maxlag)
Theta = VARMA_model.irf(maxlag)
Gamma = VARMA_model.acvf(maxlag)
C = varmapack.autocov(X3[0], maxlag)
print("\nVARMA model analysis:")
print(f"spectral radius: {rho:.6f}")
print(f"MA spectral radius: {rhoMA:.6f}")
print("Psi, Theta, Gamma, and sample autocovariance shapes:")
print(Psi.shape, Theta.shape, Gamma.shape, C.shape)
